File: src/domain/error-taxonomy.js
// ═══════════════════════════════════════════════════════════
// Taxonomia expandida de erros de pilotagem
// ═══════════════════════════════════════════════════════════
// Complementa ErrorClassifier com rótulos de detalhe pedagógico.
// Não substitui — o classificador base dá UMA principal.
// Esta taxonomia é usada pelo BOX para enriquecer análise pós-volta.
//
// Regra SPEC_MENSAGENS: frases do cockpit vêm da biblioteca controlada.
// Esta taxonomia NÃO vira mensagem de cockpit — só relatório e análise.

import { ErroTipo, ErroLabel } from './error-classifier.js';

// Rótulos de detalhe (execução/linha/corda).
export const ErroTipoExtra = Object.freeze({
  ...ErroTipo,
  PEGOU_CORDA:         'pegou-corda',
  NAO_PEGOU_CORDA:     'nao-pegou-corda',
  USOU_PISTA_TODA:     'usou-pista-toda',
  FREADA_INSTAVEL:     'freada-instavel',
  ENTRADA_SUJA:        'entrada-suja',
  SAIDA_RASGADA:       'saida-rasgada',
  TRAIL_BRAKING_OK:    'trail-braking-ok',
  CORRIGIU_LINHA:      'corrigiu-linha',
  ABORTOU_CURVA:       'abortou-curva',
  EXPLORACAO_OK:       'exploracao-ok',
});

export const ErroLabelExtra = Object.freeze({
  ...ErroLabel,
  'pegou-corda':         'Pegou a corda',
  'nao-pegou-corda':     'Não pegou corda',
  'usou-pista-toda':     'Usou pista toda',
  'freada-instavel':     'Freada instável',
  'entrada-suja':        'Entrada suja',
  'saida-rasgada':       'Saída rasgada',
  'trail-braking-ok':    'Trail braking ok',
  'corrigiu-linha':      'Corrigiu linha',
  'abortou-curva':       'Abortou curva',
  'exploracao-ok':       'Exploração ok',
});

/** Categoria de cada rótulo — pro box agrupar. */
export const CategoriaErro = Object.freeze({
  'freou-cedo':          'frenagem',
  'freou-tarde-demais':  'frenagem',
  'boa-frenagem':        'frenagem',
  'freada-instavel':     'frenagem',
  'trail-braking-ok':    'frenagem',

  'entrou-forte-demais': 'entrada',
  'entrou-timido':       'entrada',
  'entrada-suja':        'entrada',

  'pegou-corda':         'linha',
  'nao-pegou-corda':     'linha',
  'manteve-linha':       'linha',
  'usou-pista-toda':     'linha',
  'corrigiu-linha':      'linha',
  'abortou-curva':       'linha',

  'perdeu-rotacao':      'saida',
  'matou-saida':         'saida',
  'perdeu-tracao':       'saida',
  'saiu-forte':          'saida',
  'saida-rasgada':       'saida',

  'exploracao-ok':       'geral',
});

/**
 * Classifica detalhes adicionais de UM trecho a partir de
 * telemetria bruta. Recebe execution enriquecido (linha, corda,
 * estabilidade) — inputs opcionais; ausência = sem rótulo.
 */
export function classifyDetails(exec = {}) {
  const out = [];
  if (exec.aderenciaCorda != null) {
    if (exec.aderenciaCorda >= 0.8)       out.push(ErroTipoExtra.PEGOU_CORDA);
    else if (exec.aderenciaCorda <= 0.35) out.push(ErroTipoExtra.NAO_PEGOU_CORDA);
  }
  if (exec.estabilidadeFrear != null) {
    if (exec.estabilidadeFrear <= 0.4) out.push(ErroTipoExtra.FREADA_INSTAVEL);
  }
  if (exec.aberturaLinha != null) {
    if (exec.aberturaLinha >= 0.85) out.push(ErroTipoExtra.USOU_PISTA_TODA);
    else if (exec.aberturaLinha <= 0.4) out.push(ErroTipoExtra.ENTRADA_SUJA);
  }
  if (exec.abortou) out.push(ErroTipoExtra.ABORTOU_CURVA);
  if (exec.corrigiu) out.push(ErroTipoExtra.CORRIGIU_LINHA);
  if (exec.trail)    out.push(ErroTipoExtra.TRAIL_BRAKING_OK);
  return out;
}

export function labelErro(key) {
  return ErroLabelExtra[key] || key;
}

export function categoriaErro(key) {
  return CategoriaErro[key] || 'geral';
}

File: ios/p1fast-core/Sources/P1FastCore/ErrorTaxonomy.swift
// ═══════════════════════════════════════════════════════════
// ErrorTaxonomy + ToleranceFromDyno — port Swift de:
//   src/domain/error-taxonomy.js
//   src/domain/tolerance-from-dyno.js
// ═══════════════════════════════════════════════════════════
// Taxonomia: extension do ErrorClassifier (MS-1.2). 10 rótulos
// extras de DETALHE (linha/corda/estabilidade) + categoria por
// rótulo. NÃO substitui o classificador base — enriquece análise
// pós-volta no box. Frases do cockpit não vêm daqui.
//
// ToleranceFromDyno: calcula tolerance_rpm automática baseada na
// "janela útil" da curva de dyno (RPMs com potência ≥ 95% do pico).
// Resultado clampeado em [80, 250].
//
// MS-6.3 (port 2026-05-04). Camada pura.

import Foundation

// MARK: - ErrorTaxonomy (extension de ErrorClassifier)

/// 10 rótulos de detalhe pedagógico (linha/corda/estabilidade).
/// Coexistem com os 16 do `ErroTipo` — alguns trechos podem ter
/// principal (de ErroTipo) + secundárias (destes).
public enum ErroTipoExtra: String, Sendable, Codable, CaseIterable {
    case pegouCorda          = "pegou-corda"
    case naoPegouCorda       = "nao-pegou-corda"
    case usouPistaToda       = "usou-pista-toda"
    case freadaInstavel      = "freada-instavel"
    case entradaSuja         = "entrada-suja"
    case saidaRasgada        = "saida-rasgada"
    case trailBrakingOk      = "trail-braking-ok"
    case corrigiuLinha       = "corrigiu-linha"
    case abortouCurva        = "abortou-curva"
    case exploracaoOk        = "exploracao-ok"
}

public enum CategoriaErro: String, Sendable, Codable, CaseIterable {
    case frenagem
    case entrada
    case linha
    case saida
    case geral
}

public enum ErrorTaxonomy {
    /// Categoria de cada rótulo (ErroTipo + ErroTipoExtra). Paridade
    /// exata com `CategoriaErro` em `error-taxonomy.js`.
    public static func categoriaForRaw(_ raw: String) -> CategoriaErro {
        switch raw {
        // frenagem
        case "freou-cedo", "freou-tarde-demais", "boa-frenagem",
             "freada-instavel", "trail-braking-ok":
            return .frenagem
        // entrada
        case "entrou-forte-demais", "entrou-timido", "entrada-suja":
            return .entrada
        // linha
        case "pegou-corda", "nao-pegou-corda", "manteve-linha",
             "usou-pista-toda", "corrigiu-linha", "abortou-curva":
            return .linha
        // saida
        case "perdeu-rotacao", "matou-saida", "perdeu-tracao",
             "saiu-forte", "saida-rasgada":
            return .saida
        default:
            return .geral
        }
    }

    public static func categoria(_ tipo: ErroTipo) -> CategoriaErro {
        categoriaForRaw(tipo.rawValue)
    }

    public static func categoria(_ extra: ErroTipoExtra) -> CategoriaErro {
        categoriaForRaw(extra.rawValue)
    }

    /// PT-BR labels dos 10 rótulos extras.
    public static func label(_ extra: ErroTipoExtra) -> String {
        switch extra {
        case .pegouCorda:        return "Pegou a corda"
        case .naoPegouCorda:     return "Não pegou corda"
        case .usouPistaToda:     return "Usou pista toda"
        case .freadaInstavel:    return "Freada instável"
        case .entradaSuja:       return "Entrada suja"
        case .saidaRasgada:      return "Saída rasgada"
        case .trailBrakingOk:    return "Trail braking ok"
        case .corrigiuLinha:     return "Corrigiu linha"
        case .abortouCurva:      return "Abortou curva"
        case .exploracaoOk:      return "Exploração ok"
        }
    }

    /// Input enriquecido pra `classifyDetails` — todos opcionais
    /// (ausência = sem rótulo).
    public struct DetailsInput: Sendable, Equatable {
        public var aderenciaCorda: Double?    // 0..1
        public var estabilidadeFrear: Double? // 0..1
        public var aberturaLinha: Double?     // 0..1
        public var abortou: Bool
        public var corrigiu: Bool
        public var trail: Bool

        public init(
            aderenciaCorda: Double? = nil,
            estabilidadeFrear: Double? = nil,
            aberturaLinha: Double? = nil,
            abortou: Bool = false,
            corrigiu: Bool = false,
            trail: Bool = false
        ) {
            self.aderenciaCorda = aderenciaCorda
            self.estabilidadeFrear = estabilidadeFrear
            self.aberturaLinha = aberturaLinha
            self.abortou = abortou
            self.corrigiu = corrigiu
            self.trail = trail
        }
    }

    /// Classifica detalhes adicionais de UM trecho. Retorna lista
    /// (pode ser vazia). Paridade exata JS.
    public static func classifyDetails(_ exec: DetailsInput) -> [ErroTipoExtra] {
        var out: [ErroTipoExtra] = []
        if let a = exec.aderenciaCorda {
            if a >= 0.8        { out.append(.pegouCorda) }
            else if a <= 0.35  { out.append(.naoPegouCorda) }
        }
        if let e = exec.estabilidadeFrear, e <= 0.4 {
            out.append(.freadaInstavel)
        }
        if let l = exec.aberturaLinha {
            if l >= 0.85       { out.append(.usouPistaToda) }
            else if l <= 0.4   { out.append(.entradaSuja) }
        }
        if exec.abortou  { out.append(.abortouCurva) }
        if exec.corrigiu { out.append(.corrigiuLinha) }
        if exec.trail    { out.append(.trailBrakingOk) }
        return out
    }
}

// MARK: - ToleranceFromDyno

public enum ToleranceFromDyno {
    public static let peakThresholdPct: Double = 0.95
    public static let tolMin: Double = 80
    public static let tolMax: Double = 250

    public enum Error: Swift.Error, Equatable, Sendable {
        case curveTooShort
    }

    /// Calcula tolerance_rpm automática a partir da curva de dyno.
    /// Janela útil = RPMs com potência ≥ 95% do pico.
    /// tolerance_rpm = janela_util_largura × percent/100, clampeado [80, 250].
    /// Lança em curva < 3 pontos. Pico inválido ou janela < 2 pontos
    /// devolve `tolMin`.
    public static func compute(curve: [DynoPoint], percent: Double = 5) throws -> Double {
        guard curve.count >= 3 else { throw Error.curveTooShort }
        let sorted = curve.sorted { $0.rpm < $1.rpm }
        var peakP = -Double.infinity
        for p in sorted where p.powerKw.isFinite && p.powerKw > peakP {
            peakP = p.powerKw
        }
        guard peakP.isFinite, peakP > 0 else { return tolMin }
        let thr = peakP * peakThresholdPct
        let inside = sorted.filter { $0.powerKw.isFinite && $0.powerKw >= thr }
        guard inside.count >= 2 else { return tolMin }
        let lo = inside.first!.rpm
        let hi = inside.last!.rpm
        let window = max(0, hi - lo)
        let tol = (window * (percent / 100)).rounded()
        return max(tolMin, min(tolMax, tol))
    }
}
